src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -0,0 +1,350 @@
+
+        // Get the cursor position in text buffer position
+        auto cursorArgs = winrt::make_self<CursorPositionEventArgs>();
+        _CurrentCursorPositionHandlers(*this, *cursorArgs);
+        const COORD cursorPos = { gsl::narrow_cast<SHORT>(cursorArgs->CurrentPosition().X), gsl::narrow_cast<SHORT>(cursorArgs->CurrentPosition().Y) };
+
+        // Get Font Info as we use this is the pixel size for characters in the display
+        auto fontArgs = winrt::make_self<FontInfoEventArgs>();
+        _CurrentFontInfoHandlers(*this, *fontArgs);
+
+        const float fontWidth = fontArgs->FontSize().Width;
+        const float fontHeight = fontArgs->FontSize().Height;
+
+        // Convert text buffer cursor position to client coordinate position within the window
+        COORD clientCursorPos;
+        COORD screenCursorPos;
+        THROW_IF_FAILED(ShortMult(cursorPos.X, gsl::narrow<SHORT>(fontWidth), &clientCursorPos.X));
+        THROW_IF_FAILED(ShortMult(cursorPos.Y, gsl::narrow<SHORT>(fontHeight), &clientCursorPos.Y));
+
+        // Convert from client coordinate to screen coordinate by adding window position
+        THROW_IF_FAILED(ShortAdd(clientCursorPos.X, gsl::narrow_cast<SHORT>(windowBounds.X), &screenCursorPos.X));
+        THROW_IF_FAILED(ShortAdd(clientCursorPos.Y, gsl::narrow_cast<SHORT>(windowBounds.Y), &screenCursorPos.Y));
+
+        // get any offset (margin + tabs, etc..) of the control within the window
+        const auto offsetPoint = this->TransformToVisual(nullptr).TransformPoint(winrt::Windows::Foundation::Point(0, 0));
+
+        // add the margin offsets if any
+        const auto currentMargin = this->Margin();
+        THROW_IF_FAILED(ShortAdd(screenCursorPos.X, gsl::narrow_cast<SHORT>(offsetPoint.X), &screenCursorPos.X));
+        THROW_IF_FAILED(ShortAdd(screenCursorPos.Y, gsl::narrow_cast<SHORT>(offsetPoint.Y), &screenCursorPos.Y));
+
+        // Get scale factor for view
+        const double scaleFactor = DisplayInformation::GetForCurrentView().RawPixelsPerViewPixel();
+
+        // Set the selection layout bounds
+        Rect selectionRect = Rect(screenCursorPos.X, screenCursorPos.Y, 0, fontHeight);
+        request.LayoutBounds().TextBounds(ScaleRect(selectionRect, scaleFactor));
+
+        // Set the control bounds of the whole control
+        Rect controlRect = Rect(screenCursorPos.X, screenCursorPos.Y, 0, fontHeight);
+        request.LayoutBounds().ControlBounds(ScaleRect(controlRect, scaleFactor));
+
+        // position textblock to cursor position
+        _canvas.SetLeft(_textBlock, clientCursorPos.X);
+        _canvas.SetTop(_textBlock, static_cast<double>(clientCursorPos.Y));
+
+        // width is cursor to end of canvas
+        _textBlock.Width(200); // TODO GitHub #3640: Determine proper Width
+        _textBlock.Height(fontHeight);
+
+        // calculate FontSize in pixels from DIPs
+        const float logicalDpi = DisplayInformation::GetForCurrentView().LogicalDpi();

```

Comment:
no no no this actually _has to be 96.0_. This renders at half size on a 192-dpi display.

The reason is: we're converting to _device-independent pixels_, but using the real DPI gives us _device-dependent pixels_ :grin:

Reply:
(i just tested it out:

![image](https://user-images.githubusercontent.com/14316954/69380733-f3d1d000-0c67-11ea-82b5-2286a0552ff7.png)


)

Reply:
oh no

Reply:
Whoops, was trying to be too smart.  Fixed!